### ISSUE TITLE
Merge using comments

### ISSUE THREAD

**Issue by rarkins** [renovatebot/renovate#1355]
Scenario:

- Repository has 5 Renovate PRs
- All 5 pass tests, and look OK to a reviewer
- Reviewer wants to merge them all "right away", but:
- To be safest, you should merge one, then rebase/retest the rest, then merge, then rebase/retest, etc

The result is that a reviewer/merger needs to keep checking back over a period of perhaps hours as they wait for tests to finish.

Solution:

- users with write permissions can comment on each with "!merge"
- Renovate will automatically merge them one by one, as it is able - allowing for rebases, retesting, etc

Inspiration: https://github.com/uber-workflow/probot-app-merge-pr

**Comment by rarkins** [comment]
To the Algolia team e.g. @samouss and @Haroenv, I often see you merging multiple PRs close together. Would this approach save you time? e.g. you manually review the PRs in a batch and add a comment "!merge" to each that is OK. Renovate will then merge them as fast as is possible allowing for any necessary conflict resolving, rebasing, retesting, etc.

**Comment by Haroenv** [comment]
I'd like it more if a review would do this, rather than `!merge`, is that possible?

**Comment by rarkins** [comment]
It's possible today, although perhaps with some warning.

First of all, how that would work:
- You would configure Renovate with automerge enabled, perhaps even for all PRs including major.
- You would enable branch protection such that Reviews are required before merging. This will "block" automerge by default
- If a Renovate PR is reviewed and other tests have passed, then GitHub will turn it green. This will allow Renovate to automerge it

The warning is that maybe:
- The person reviewing it does not *mean* "merge this!". They maybe mean "looks ok to me" but assume someone else will look as well
- Maybe the person reviewing should not be authorised to approve a merge

**Comment by Haroenv** [comment]
Those are good points, is there some way to check if reviewer can merge?

**Comment by rarkins** [comment]
If you have restricted who can merge to master in the repo (e.g. in branch protection settings you have listed individuals or teams with merge rights) then unfortunately Renovate cannot merge. This is because of a GitHub bug/limitation where they don't offer any way to add an *app* (bot) to that list so all apps get blocked from merging if you enable that type of branch protection.

Otherwise, if you just mean users with write access to the repository vs read-only, then we can probably look that up. Please confirm if that's what you mean or if it's something else.

**Comment by rarkins** [comment]
Maybe.. we could also add a new config option to renovate's config that lets you define which users should merge if they review/approve the PR.

**Comment by destroyerofbuilds** [comment]
I like the idea of using built-in functionality (reviewers) of the GitHub interface for this workflow.

(I use the [merge when pipeline succeeds](https://docs.gitlab.com/ce/user/project/merge_requests/merge_when_pipeline_succeeds.html) feature in GitLab quite often to help queue up merge requests when several come in at once)

**Comment by rarkins** [comment]
Closed in favour of #3172

**Comment by rarkins** [comment]
Reopening to add this comment by @azu in #3172:

Merge via comment is useful for me.
GitHub/GitLab supports commenting via email and I can merge the dependabot's PR from email client like gmail.

- [Reply to Comments from Email - The GitHub Blog](https://github.blog/2011-03-10-reply-to-comments-from-email/)

UseCase: renovatebot/dependabot create same Pull Request to multiple repositories at same time.
It is hard that open each PRs and click merge button.
Instead of it, I can just reply `@dependabot merge` from email client.

<img width="1067" alt="image" src="https://user-images.githubusercontent.com/19714/102346908-fa558700-3fe2-11eb-9ff7-c971dd23d4a0.png">

checkbox approach may not cover this case.

**Comment by rarkins** [comment]
GitHub's native automerge feature is a viable and someway better workaround for this

**Comment by anantakrishna** [comment]
> GitHub's native automerge feature is a viable and someway better workaround for this

Can you please give a link? I cannot find this feature by googling.

**Comment by rarkins** [comment]
Google for "GitHub automerge"

**Comment by anantakrishna** [comment]
Thanks, and excuse me for a bit of laziness. I was searching for “GitHub's native automerge feature”, and mostly only Dependabot-related articles came up.

Here is the link for future strangers: https://docs.github.com/en/github/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/automatically-merging-a-pull-request

**Comment by fastman** [comment]
> GitHub's native automerge feature is a viable and someway better workaround for this

Yeah, but the world does not end on Github :-) Still the feature is desired 👍

**Comment by rushilsrivastava** [comment]
> Which platform/hosting mode are you using?

We use the GitHub App.

**Comment by rarkins** [comment]
Rebasing should happen automatically in the GitHub app without requiring manual intervention. Please create a separate discussion topic if you're not seeing that and can provide some logs or other evidence to help pinpoint it

**Comment by rarkins** [comment]
Poetry may be slow and that's not Renovate's fault, but the following should be false:

> (Between 10-15 minutes if manually triggered, 25-30 minutes if automatic)

Automatic rebasing should happen as fast/faster than you can achieve with manual, because the merging of any Renovate PR should trigger a priority job in the queue.

I would have hoped that if you enable GitHub automerge in each PR and just leave them you should see them rebased until they are all merged (or failing tests).

**Comment by Inok** [comment]
Such "semi-auto" merge will be useful. In my case, I don't want to merge everything manually, but I want to review the changelog of the package being updated and changes in my source code before merging them.

Maybe it's possible to add an option that enables such "semi-auto" flow? For example, by adding a new checkbox to PRs? (like `force rebase` checkbox)
If it's checked, Renovate will try to rebase/merge the PR automatically. If it's not - Renovate will wait for confirmation.

So `automerge` option may be changed from a boolean flag to something like:
- `enabled` - current `automerge=true`
- `suggested` - disabled until the checkbox is checked, then enabled
- `disabled` - current `automerge=false`

Or simply add `automergeType=pr-with-confirmation`, which enables such behavior.

**Comment by rarkins** [comment]
`automergeType=checkbox` might make most sense

**Comment by HonkingGoose** [comment]
@rarkins said they want to do the linked issue first: [^quote]

- #20376

I'll mark this issue `status:blocked`. After we've tried the checkbox-approach, we'll decide if we want to work on this issue. 😉 

[^quote]: https://github.com/renovatebot/renovate/discussions/20166#discussioncomment-4852270

**Comment by HonkingGoose** [comment]
Related issue:

- #20819

**Comment by HonkingGoose** [comment]
## Summary

How about prioritizing:

- #5573

And closing these issues/discussions as won't do:

- #1355
- #24911

That way the bulk of users have a good way to speed up the "merge multiple Renovate PRs" workflow.

### Why we opened this issue

This is the problem we wanted to solve by making a new "merge using comments" feature:

> Scenario:
> 
> - Repository has 5 Renovate PRs
> - All 5 pass tests, and look OK to a reviewer
> -Reviewer wants to merge them all "right away", but:
> - To be safest, you should merge one, then rebase/retest the rest, then merge, then rebase/retest, etc
> 
> The result is that a reviewer/merger needs to keep checking back over a period of perhaps hours as they wait for tests to finish.

### Changes since the issue was created

#### GitHub Merge Queue

GitHub now has Merge Queue, [^footnote-merge-queue] where you can approve a PR, put it in the Queue, and let GitHub merge when the tests pass. GitHub will make sure the combination of PRs passes the tests. You can use GitHub Merge Queue with Renovate. [^footnote-renovate-docs]

#### GitLab has Merge Trains (we don't support that yet)

- #5573

#### We're going to a checkbox-based system?

I think, in general, we're going towards a checkbox-based system? See:

- #20376
- #29712
- #21719 (checkbox to force open pending version due to `internalChecksFilter=strict`)

### Difficult to verify comment is allowed to merge

If we want to keep this issue open, we still need to decide how to get Renovate to only accept merge comments from allowed users:

- #24911

[^footnote-merge-queue]: [GitHub Docs, managing a merge queue](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/managing-a-merge-queue)
[^footnote-renovate-docs]: [Renovate docs, GitHub Merge Queue](https://docs.renovatebot.com/key-concepts/automerge/#github-merge-queue)
